### PR TITLE
Initial app

### PR DESCRIPTION
**Pull request by cbartz**
Applicable spec: n/a

### Overview

Checkin code (with the latest version from https://github.com/canonical/github-runner-operator/pull/351). Adjustments had to be made to remove the charm_state dependencies (these dependencies have been moved to the `types_` package). There is also some code duplication with the code in https://github.com/canonical/github-runner-operator (`github_client`, `repo_policy_compliance_client`, `errors`, `utilities`). One reason for this is that the LXDRunnerManager requires this functionality, but is not currently intended to be outsourced.

### Rationale

The primary reason for the externalisation is mainly to remove the charm_state dependencies, as the WIP reactive code will spawn runners in a different process using a script, and serialising the charm state and passing as parameter to a script has been identified as a bad smell. The deeper cause is that the github runner manager can be seen as a separate application, and https://github.com/canonical/github-runner-operator has too much responsibility.

### Checklist

- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The documentation is generated using `src-docs`
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1 +1,6 @@
-ops >= 2.2.0
+fabric >=3,<4
+ghapi
+jinja2
+kombu==5.4.0
+openstacksdk>=3,<4
+pydantic < 2
```
Comment (jdkandersson) [inline requirements.txt]: new line

File: src/github_runner_manager/errors.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+class ConfigurationError(Exception):
+    """Error for juju configuration."""
+
+
+class MissingMongoDBError(Exception):
+    """Error for missing integration data."""
+
+
+class LxdError(Exception):
```
Comment (jdkandersson) [inline src/github_runner_manager/errors.py]: Check this error

File: src/github_runner_manager/errors.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        self.return_code = return_code
+        self.stdout = stdout
+        self.stderr = stderr
+
+
+class IssueMetricEventError(Exception):
+    """Represents an error when issuing a metric event."""
+
+
+class LogrotateSetupError(Exception):
```
Comment (jdkandersson) [inline src/github_runner_manager/errors.py]: Check this?

File: src/github_runner_manager/errors.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    """Represents an error raised when logrotate cannot be setup."""
+
+
+class MetricsStorageError(Exception):
+    """Base class for all metrics storage errors."""
+
+
+class SharedFilesystemError(MetricsStorageError):
```
Comment (jdkandersson) [inline src/github_runner_manager/errors.py]: Check this one

File: src/github_runner_manager/reactive/runner_manager.py
Comment (jdkandersson) [inline src/github_runner_manager/reactive/runner_manager.py:1]: We should probably move the script here

File: src/github_runner_manager/types_/general.py
Comment (jdkandersson) [inline src/github_runner_manager/types_/general.py:1]: Perhaps move them to the __init__ file

**Review by jdkandersson — DISMISSED** [review]

**Review by jdkandersson — DISMISSED** [review]

File: .github/pull_request_template.yaml [hunk 5]
```diff
@@ -8,25 +8,19 @@ Applicable spec: <link>
...
 
 ### Checklist
 
-- [ ] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
 - [ ] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
-- [ ] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
 - [ ] The documentation is generated using `src-docs`
-- [ ] The documentation for charmhub is updated
 - [ ] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)
+- [ ] The application version in `pyproject.toml` is incremented
```
Comment (yanksyoon) [inline .github/pull_request_template.yaml]:
```suggestion
- [ ] The library version in `pyproject.toml` is incremented
```

File: pyproject.toml [hunk 6]
```diff
@@ -1,3 +1,33 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+[project]
+name = "github-runner-manager"
+version = "0.2.0"
```
Comment (yanksyoon) [inline pyproject.toml]:
```suggestion
version = "0.1.0"
```

File: pyproject.toml [hunk 7]
```diff
@@ -6,14 +36,39 @@ skips = ["*/*test.py", "*/test_*.py", "*tests/*.py"]
 # Testing tools configuration
 [tool.coverage.run]
 branch = true
+omit = [
+    # These are covered by `tests/integration/test_runner_manager_openstack.py` in the github-runner-operator.
+    "src/github_runner_manager/openstack_cloud/openstack_cloud.py",
+    "src/github_runner_manager/openstack_cloud/openstack_runner_manager.py",
+    # Contains interface for calling repo policy compliance service. Tested in integration test
+    # and end to end tests.
+    "src/github_runner_manager/repo_policy_compliance_client.py",
```
Comment (yanksyoon) [inline pyproject.toml:45]: I think we can remove these and slowly work towards integration test coverage :)

File: requirements.txt
Comment (yanksyoon) [inline requirements.txt:1]: Out of scope for this PR, as a reminder to myself: start pinning the requirements

**Review by yanksyoon — DISMISSED** [review]

**Comment by yanksyoon** [comment]
Just some comments for myself, thank you for the PR!

**Review by jdkandersson — APPROVED** [review]

**Review by yanksyoon — APPROVED** [review]

**Pull request merged**